File: Sorting/SortingAlgorithmsVisualisation/src/main/kotlin/MVC/Sorter/BubbleSorter.kt
package MVC.Sorter

class BubbleSorter : Sorter() {
    override fun sortAndReturnIndex(arrayToSort: Array<Int>, currentIndex: Int) : Int {
        return if (currentIndex < arrayToSort.size - 1) {
            if (arrayToSort[currentIndex] > arrayToSort[currentIndex + 1]) {
                swap(arrayToSort, currentIndex, currentIndex + 1)
            }
            currentIndex + 1
        } else {
            0
        }
    }
}

File: src/Sorting/Sorters/bubbleSorter.cpp
#include "bubbleSorter.hpp"
#include "../../Utils/Timer/timer.hpp"

BubbleSorter::BubbleSorter()
	: Sorter(SortingAlgorithm::Bubble)
{
}

void BubbleSorter::sort()
{
	Timer timer;
	timer.reset();

	while (true)
	{

		for (m_index = 1; m_index < m_array.size(); m_index++)
		{
			if (m_array[m_index] < m_array[m_index - 1])
			{
				m_arrayMutex.lock();
				{
					std::swap(m_array[m_index], m_array[m_index - 1]);
				}
				m_arrayMutex.unlock();
			}

			if (m_stopSorting)
			{
				return;
			}

			while (!m_continueSorting.load());
			applyDelay(timer.getTimeInSeconds());
			timer.reset();

		}

		if (isArraySorted())
		{
			m_finishedSorting = true;
			m_isSorting = false;
			break;
		}
	}
}

int BubbleSorter::getIndex() const
{
	return m_index;
}
